Assignment2Q2.java

```diff
@@ -0,0 +1,112 @@
+import java.util.LinkedList;
+
+/**
+ * Created by Alexandra on 23.05.2017.
+ */
+public class Assignment2Q2 {
+
+    public static class TreeNode <T> {
+        private TreeNode<T> left;
+        private TreeNode<T> right;
+        private T value;
+
+        public TreeNode(TreeNode<T> left, TreeNode<T> right, T value) {
+            if (value == null) {
+                throw new IllegalArgumentException();
+            }
+            this.left = left;
+            this.right = right;
+            this.value = value;
+        }
+
+        public TreeNode<T> getLeft() {
+            return left;
+        }
+
+        public TreeNode<T> getRight() {
+            return right;
+        }
+
+        public T getValue() {
+            return value;
+        }
+
+        public String toString() {
+            return value.toString() + '\t';
+        }
+
+        public T lowestCommonAncestor(T value1, T value2) {
+            // since there is only one path from a node to the root

```

Comment:
This is a nice efficient algorithm, well done. Depth first search with an intermediate data structure. Very clear algorithmic and memory efficiency.